lib/packetgen/header/http/request.rb

```diff
@@ -0,0 +1,92 @@
+# This file is part of PacketGen
+# See https://github.com/sdaubert/packetgen for more informations
+# Copyright (C) 2016 Sylvain Daubert <sylvain.daubert@laposte.net>
+# This program is published under MIT license.
+
+module PacketGen
+  module Header
+    module HTTP
+      # An HTTP/1.1 Request packet consits of:
+      # * the http method ({Types::String}).
+      # * the path ({Types::String}).
+      # * the version ({Types::String}).
+      # * associated http headers ({Types::String}).
+      #
+      # == Create a HTTP Request header
+      #   # standalone
+      #   http_rqst = PacketGen::Header::HTTP::Request.new
+      #   # in a packet
+      #   pkt = PacketGen.gen("IP").add("TCP").add("HTTP::Request")
+      #   # access to HTTP Request header
+      #   pkt.http_request # => PacketGen::Header::HTTP::Request
+      #
+      # == HTTP Request attributes
+      #	  http_rqst.version = "HTTP/1.1"
+      #	  http_rqst.method = "GET"
+      #	  http_rqst.path = "/meow.html"
+      #   http_rqst.headers = "Host: tcpdump.org"     # string or
+      #	  http_rqst.headers = { host: "tcpdump.org" } # even a hash
+      #
+      # @author Kent 'picat' Gruber
+      class Request < Base
+        # @!attribute method
+        #   @return [Types::String]
+        define_field :method,  Types::String
+        # @!attribute path 
+        #   @return [Types::String]
+        define_field :path,    Types::String
+        # @!attribute version 
+        #   @return [Types::String]
+        define_field :version, Types::String, default: "HTTP/1.1"
+        # @!attribute headers
+        #   associated http/1.1 headers
+        #   @return [Types::HTTPHeaders]
+        define_field :headers, Types::HTTPHeaders 
+
+        # @param [Hash] options
+        # @option options [String] :method
+        # @option options [String] :path
+        # @option options [String] :version
+        # @option options [Hash]   :headers
+        def initialize(options={})
+          super(options)
+          self.headers ||= options[:headers]
+        end
+
+        # Give it a pretty name, the only way I know how.
+        # @return [String]
+        def protocol_name
+          "HTTP_Request"
+        end
```

Comment:
should return "HTTP::Request" : this is the name that should be used in PacketGen.gen and #add

Reply:
The `#protocol_name` method also seems to be linked to the magic that handles how the method name is generated like `pkt.http_request` or `pkt.http_response` -- otherwise with just `HTTP::Response` it will be `pkt.response` and not the full thing I want. 

Is there a better way to do this?

Reply:
Yes, there is!

Do nothing: `#protocol_name` is already defined by Headers::Base and works properly. It is used by `Header::Base#method_name` which really defines header accessor on Packet.

Reply:
The reason I made this change was becuase I was experiencing the following behavior:

```ruby
    184: def protocol_name
 => 185:   binding.pry
    186:   self.class.to_s.sub(/.*::/, '')
    187: end

pg(#<PacketGen::Header::HTTP::Request>)> self
=> ---- PacketGen::Header::HTTP::Request --------------------------------
          String       method: "GET"
          String         path: "/manpages/pcap.3pcap.html"
          String      version: "HTTP/1.1"
         Headers      headers: {:host=>"www.tcpdump.org", :user_agent=>"android-async-http/1.4.4 (http://loopj.com/android-async-http)", :accept=>"*/*"}

pg(#<PacketGen::Header::HTTP::Request>)> self.class
=> PacketGen::Header::HTTP::Request
pg(#<PacketGen::Header::HTTP::Request>)> self.class.to_s
=> "PacketGen::Header::HTTP::Request"
pg(#<PacketGen::Header::HTTP::Request>)> self.class.to_s.sub(/.*::/, '')
=> "Request"
pg(#<PacketGen::Header::HTTP::Request>)> 
```

I think I'd rather have the method name to be `http_request` I think instead of simply `request`.

Reply:
To simplify what I mean by this: `self.class.to_s.sub(/.*::/, '')` will turn `PacketGen::Header::HTTP::Request` to `"Request"`

I think it's nicer to have the method name `#http_request` which is why I made the  `#protocol_name` method have `"HTTP_Request"` as the output.

While I could still be wrong for doing that, that is why I did it.

Reply:
🤔  Just a thought:

```ruby
def protocol_name
  self.class.to_s.split("::Header::").last.sub(/::/, '_')
end
```

Could that maybe help fix this what I want for the `#protocol_name` while still working for other PacketGen headers?